File: 340-longest-substring-with-at-most-k-distinct-characters/longest-substring-with-at-most-k-distinct-characters.py
class Solution:
    def lengthOfLongestSubstringKDistinct(self, s: str, k: int) -> int:
        st=0
        e=0
        ans=0
        d={}
        while e<len(s):
            d[s[e]]=d.get(s[e],0)+1
            if len(d)<=k:
                ans=max(ans,e-st+1)
                e=e+1
            else:
                while len(d)>k and st<=e:
                    d[s[st]]=d[s[st]]-1
                    if d[s[st]]==0:
                        del d[s[st]]
                    st=st+1
                # if len(d)<=k:
                #     ans=max(ans,e-st+1)
                e=e+1
        return ans
        

File: 340-longest-substring-with-at-most-k-distinct-characters/longest-substring-with-at-most-k-distinct-characters.java
class Solution {
    public int lengthOfLongestSubstringKDistinct(String s, int k) {
        int left=0;
        int right=0;
        int length=s.length();
        HashMap<Character,Integer> map=new HashMap<>();
        int maxi=0;
        while(right<length){
            map.put(s.charAt(right),map.getOrDefault(s.charAt(right),0)+1);
            while(map.size()>k){
               map.put(s.charAt(left),map.get(s.charAt(left))-1);
               if(map.get(s.charAt(left))==0){
                    map.remove(s.charAt(left));
               } 
               left=left+1;
            }
            if(map.size()<=k){
                maxi=Math.max(maxi,right-left+1);
            }
            right++;
        }
        return maxi;
    }
}